cpp/src/streaming/cudf/partition.cpp

```diff
@@ -0,0 +1,117 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <memory>
+
+#include <cudf/partitioning.hpp>
+
+#include <rapidsmpf/integrations/cudf/partition.hpp>
+#include <rapidsmpf/streaming/cudf/partition.hpp>
+
+namespace rapidsmpf::streaming::node {
+
+Node partition_and_pack(
+    std::shared_ptr<Context> ctx,
+    SharedChannel<TableChunk> ch_in,
+    SharedChannel<PartitionMapChunk> ch_out,
+    std::vector<cudf::size_type> columns_to_hash,
+    int num_partitions,
+    cudf::hash_id hash_function,
+    uint32_t seed
+) {
+    ShutdownAtExit c{ch_in, ch_out};
+
+    co_await ctx->executor()->schedule();
+    while (true) {
+        auto table = co_await ch_in->receive_or(nullptr);
+        if (table == nullptr) {
+            break;
+        }
+        auto reservation = ctx->reserve_and_spill(table->make_available_cost());
+        auto tbl = table->make_available(reservation, table->stream(), ctx->br());
+
+        PartitionMapChunk partition_map{
+            .sequence_number = tbl.sequence_number(),
+            .data = rapidsmpf::partition_and_pack(
+                tbl.table_view(),
+                columns_to_hash,
+                num_partitions,
+                hash_function,
+                seed,
+                table->stream(),
+                ctx->br(),
+                ctx->statistics()
+            ),
+            .stream = tbl.stream()
+        };
```

Comment:
Nit. I'd prefer adding a constructor to the struct TBH. This is a bit too verbose ;-)